src/Microsoft.ML.Featurizers/Common.cs

```diff
@@ -0,0 +1,204 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Text;
+using Microsoft.Win32.SafeHandles;
+
+namespace Microsoft.ML.Featurizers
+{
+    #region Native Function Declarations
+
+    #endregion
+
+    internal enum FitResult : byte
+    {
+        Complete = 1, Continue, ResetAndContinue
+    }
+
+    // Not all these types are currently supported. This is so the ordering will allign with the native code.
+    internal enum TypeId : uint
+    {
+        String = 1,
+        SByte,
+        Short,
+        Int,
+        Long,
+        Byte,
+        UShort,
+        UInt,
+        ULong,
+        Float16,
+        Float32,
+        Double,
+        Complex64,
+        Complex128,
+        BFloat16,
+        Bool,
+        Timepoint,
+        Duration,
+
+        LastStaticValue,
+        Tensor = 0x1001 | LastStaticValue + 1,
+        SparseTensor = 0x1001 | LastStaticValue + 2,
+        Tabular = 0x1001 | LastStaticValue + 3,
+        Nullable = 0x1001 | LastStaticValue + 4,
+        Vector = 0x1001 | LastStaticValue + 5,
+        MapId = 0x1002 | LastStaticValue + 6
```

Comment:
>MapId = 0x1002 | LastStaticValue + 6 [](start = 8, length = 36)

please add comments for this

Reply:
Why does this add `LastStaticValue`? That shifts all of the lower items by 1 everytime a new type is inserted.

Why the math? Seems more simple to directly map to the values.

Why is the base of the last one `0x1002`? That skips a value.

Reply:
All of these are taken directly as is from the Native code base. They are here just so we have a correct mapping on our side.

`LastStaticValue` is supposed to represent the last stand alone type. All the types that follow after it are compound types. There is probably a more appropriate name, but since I want to keep this identical to the Native code I left it as is.

The math is there just to keep it identical to the native code. I could directly map the values, but then I would have to map the values every time I need to check if it has changed in the native code. Leaving it like this makes comparisons much easier.

The number at the end of `0x100N` represents how many types come after it that are associated to it. So `Tabular`, `Nullable`, and `Vector`, end in `1` meaning only 1 type follows it. The `Map` ends in `2` because we need types for the Keys and for the Values.

---
In reply to: [348115118](https://github.com/dotnet/machinelearning/pull/4413#discussion_r348115118) [](ancestors = 348115118)